File: Realize/6Week/PracticeTree2.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self, arr):
        self.length = len(arr)
        nodes = [Node(data) for data in arr]

        for i in range(self.length):
            left = 2 * i + 1
            right = 2 * i + 2
            if left < self.length:
                nodes[i].left = nodes[left]
            if right < self.length:
                nodes[i].right = nodes[right]

        self.head = nodes[0]

    def pre_order(self, node):
        if not node:
            return
        print(node.data, end=' ')
        self.pre_order(node.left)
        self.pre_order(node.right)

    def in_order(self, node):
        if not node:
            return
        self.in_order(node.left)
        print(node.data, end=' ')
        self.in_order(node.right)

    def post_order(self, node):
        if not node:
            return
        self.post_order(node.left)
        self.post_order(node.right)
        print(node.data, end=' ')

    def level_order(self, node):
        if not node:
            return
        queue = deque()
        queue.append(node)

        while queue:
            cur = queue.popleft()
            print(cur.data, end=' ')
            if cur.left:
                queue.append(cur.left)
            if cur.right:
                queue.append(cur.right)

if __name__ == "__main__":
    arr = [chr(ord('A') + i) for i in range(10)]
    bt = BinaryTree(arr)

    print("== Preorder ==")
    bt.pre_order(bt.head)
    print()

    print("== Inorder ==")
    bt.in_order(bt.head)
    print()

    print("== Postorder ==")
    bt.post_order(bt.head)
    print()

    print("== Levelorder ==")
    bt.level_order(bt.head)
    print()

File: Realize/6Week/PracticeTree2.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

class Node {
public:
    char data;
    Node* left;
    Node* right;

    Node(char data) {
        this->data = data;
        this->left = nullptr;
        this->right = nullptr;
    }
};

class BinaryTree {
public:
    Node* head;
    int length;

    BinaryTree(vector<char>& arr) {
        length = arr.size();
        vector<Node*> nodes(length);
        for (int i = 0; i < length; i++) {
            nodes[i] = new Node(arr[i]);
        }
        for (int i = 0; i < length; i++) {
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            if (left < length) nodes[i]->left = nodes[left];
            if (right < length) nodes[i]->right = nodes[right];
        }
        head = nodes[0];
    }

    void preOrder(Node* node) {
        if (!node) return;
        cout << node->data << " ";
        preOrder(node->left);
        preOrder(node->right);
    }

    void inOrder(Node* node) {
        if (!node) return;
        inOrder(node->left);
        cout << node->data << " ";
        inOrder(node->right);
    }

    void postOrder(Node* node) {
        if (!node) return;
        postOrder(node->left);
        postOrder(node->right);
        cout << node->data << " ";
    }

    void levelOrder(Node* node) {
        if (!node) return;
        queue<Node*> q;
        q.push(node);

        while (!q.empty()) {
            Node* cur = q.front();
            q.pop();
            cout << cur->data << " ";
            if (cur->left) q.push(cur->left);
            if (cur->right) q.push(cur->right);
        }
    }
};

int main() {
    vector<char> arr(10);
    for (int i = 0; i < 10; i++) {
        arr[i] = 'A' + i;
    }

    BinaryTree bt(arr);

    cout << "== Preorder ==" << endl;
    bt.preOrder(bt.head);
    cout << endl;

    cout << "== Inorder ==" << endl;
    bt.inOrder(bt.head);
    cout << endl;

    cout << "== Postorder ==" << endl;
    bt.postOrder(bt.head);
    cout << endl;

    cout << "== Levelorder ==" << endl;
    bt.levelOrder(bt.head);
    cout << endl;

    return 0;
}
